## 记一次异常深度区域的查找

```bash
sambamba markdup -t 4 --tmpdir='./' -r  123562-1055-MITO-sorted.bam \
       123562-1055-MITO-sorted.bam.rmdup.bam

samtools view 123562-1055-MITO-sorted.bam chrM:16502-16525 | wc -l
# 23143870
samtools view 123562-1055-MITO-sorted.bam chrM:16527-16529 | wc -l
# 21745885
samtools view 123562-1055-MITO-sorted.bam.rmdup.bam chrM:16502-16525 | wc -l
# 13529616
samtools view 123562-1055-MITO-sorted.bam.rmdup.bam chrM:16527-16529 | wc -l
# 12155725

bedtools genomecov -d -ibam 123560-1055-MITO-sorted.bam -g \
       /data1/user/chenjun/opengk-gene/mito/mito.fa >123560-1055-MITO-sorted.bam.genome_cov.txt
bedtools genomecov -d -ibam 123562-1055-MITO-sorted.bam -g \
       /data1/user/chenjun/opengk-gene/mito/mito.fa >123562-1055-MITO-sorted.bam.genome_cov.txt 
paste  123560-1055-MITO-sorted.bam.genome_cov.txt  123562-1055-MITO-sorted.bam.genome_cov.txt >  genome_cov.txt # （异常截图见下方）

bedtools getfasta -fi /data1/user/chenjun/opengk-gene/mito/mito.fa \
        -bed chrM:16485-16546
# AACTGTATCCGACATCTGGTTCCTACTTCAGGGTCATAAAGCCTAAATAGCCCACACGTTC
```

```tsv
chrM	16474	100443	chrM	16474	36225
chrM	16475	101093	chrM	16475	36098
chrM	16476	101774	chrM	16476	35933
chrM	16477	102573	chrM	16477	56957
chrM	16478	103347	chrM	16478	57127
chrM	16479	104205	chrM	16479	68847
chrM	16480	105075	chrM	16480	68941
chrM	16481	106304	chrM	16481	69205
chrM	16482	106953	chrM	16482	69387
chrM	16483	107823	chrM	16483	70072
chrM	16484	108646	chrM	16484	214888
chrM	16485	109639	chrM	16485	1.66205e+07
chrM	16486	110331	chrM	16486	1.70518e+07
chrM	16487	110949	chrM	16487	1.85295e+07
chrM	16488	111513	chrM	16488	1.86664e+07
chrM	16489	112156	chrM	16489	1.88046e+07
chrM	16490	112853	chrM	16490	1.89295e+07
chrM	16491	113487	chrM	16491	1.90498e+07
chrM	16492	114363	chrM	16492	1.9159e+07
chrM	16493	115053	chrM	16493	1.92661e+07
...
chrM	16536	140508	chrM	16536	2.07705e+07
chrM	16537	141072	chrM	16537	2.06738e+07
chrM	16538	141400	chrM	16538	2.05514e+07
chrM	16539	141799	chrM	16539	1.93963e+07
chrM	16540	142290	chrM	16540	1.92223e+07
chrM	16541	142248	chrM	16541	1.90806e+07
chrM	16542	142134	chrM	16542	1.8874e+07
chrM	16543	142054	chrM	16543	1.86956e+07
chrM	16544	142016	chrM	16544	1.85601e+07
chrM	16545	141944	chrM	16545	1.84258e+07
chrM	16546	141824	chrM	16546	1.67963e+07
chrM	16547	141692	chrM	16547	131332
chrM	16548	141592	chrM	16548	43052
chrM	16549	141420	chrM	16549	42269
chrM	16550	141271	chrM	16550	42035
chrM	16551	141151	chrM	16551	41788
chrM	16552	141067	chrM	16552	23172
chrM	16553	141005	chrM	16553	23002
chrM	16554	140869	chrM	16554	22844
chrM	16555	140671	chrM	16555	22621
chrM	16556	140633	chrM	16556	22572
chrM	16557	139675	chrM	16557	22534
```


bamdst统计：
picard 和 sambda 重复标记对比，sambda速度明显快很多
```tsv
[Total] Read1(rmdup)    31390695    |                                          [Total] Read1(rmdup)    31390694    
[Total] Read2(rmdup)    31375959    |                                          [Total] Read2(rmdup)    31375960    
```

```txt
## The file was created by bamdst                                           |  ## The file was created by bamdst
  ## Version : 1.0.9                                                          |  ## Version : 1.0.9
  ## Files : 124292-9100-WES-sorted_duplicates.bam                            |  ## Files :                                                                                                                      
                                 [Total] Raw Reads (All reads)    84433250    |                                 [Total] Raw Reads (All reads)    84433250
                                         [Total] QC Fail reads    0           |                                         [Total] QC Fail reads    0
                                          [Total] Raw Data(Mb)    12635.98    |                                          [Total] Raw Data(Mb)    12635.98
                                          [Total] Paired Reads    84356242    |                                          [Total] Paired Reads    84356242
                                          [Total] Mapped Reads    84249671    |                                          [Total] Mapped Reads    84249671
                              [Total] Fraction of Mapped Reads    99.78%      |                              [Total] Fraction of Mapped Reads    99.78%
+ +--  4 行: [Total] Mapped Data(Mb) 12616.60---------------------------------|+ +--  4 行: [Total] Mapped Data(Mb) 12616.60-------------------------------------------------------------------------------------
                                  [Total] Read and mate paired    84199754    |                                  [Total] Read and mate paired    84199754
                      [Total] Fraction of Read and mate paired    99.72%      |                      [Total] Fraction of Read and mate paired    99.72%
                                            [Total] Singletons    49917       |                                            [Total] Singletons    49917
                         [Total] Read and mate map to diff chr    276376      |                         [Total] Read and mate map to diff chr    276376
                                                 [Total] Read1    42178121    |                                                 [Total] Read1    42178121
                                                 [Total] Read2    42178121    |                                                 [Total] Read2    42178121
                                          [Total] Read1(rmdup)    31390695    |                                          [Total] Read1(rmdup)    42143445                                                        
                                          [Total] Read2(rmdup)    31375959    |                                          [Total] Read2(rmdup)    42106226                                                        
                                  [Total] forward strand reads    42120706    |                                  [Total] forward strand reads    42120706
                                 [Total] backward strand reads    42128965    |                                 [Total] backward strand reads    42128965
                                   [Total] PCR duplicate reads    21483017    |                                   [Total] PCR duplicate reads    0                                                               
                       [Total] Fraction of PCR duplicate reads    25.50%      |                       [Total] Fraction of PCR duplicate reads    0.00%                                                           
                              [Total] Map quality cutoff value    20          |                              [Total] Map quality cutoff value    20
                         [Total] MapQuality above cutoff reads    82000310    |                         [Total] MapQuality above cutoff reads    82000310
                   [Total] Fraction of MapQ reads in all reads    97.12%      |                   [Total] Fraction of MapQ reads in all reads    97.12%
                [Total] Fraction of MapQ reads in mapped reads    97.33%      |                [Total] Fraction of MapQ reads in mapped reads    97.33%
                                         [Target] Target Reads    73727828    |                                         [Target] Target Reads    73727828
                [Target] Fraction of Target Reads in all reads    87.32%      |                [Target] Fraction of Target Reads in all reads    87.32%
             [Target] Fraction of Target Reads in mapped reads    87.51%      |             [Target] Fraction of Target Reads in mapped reads    87.51%
                                      [Target] Target Data(Mb)    8729.60     |                                      [Target] Target Data(Mb)    8729.60
                                [Target] Target Data Rmdup(Mb)    6279.12     |                                [Target] Target Data Rmdup(Mb)    8510.53
                  [Target] Fraction of Target Data in all data    69.09%      |                  [Target] Fraction of Target Data in all data    69.09%
               [Target] Fraction of Target Data in mapped data    69.19%      |               [Target] Fraction of Target Data in mapped data    69.19%
                                        [Target] Len of region    60213773    |                                        [Target] Len of region    60213773
                                        [Target] Average depth    144.98      |                                        [Target] Average depth    144.98
                                 [Target] Average depth(rmdup)    104.28      |                                 [Target] Average depth(rmdup)    141.34
                                       [Target] Coverage (>0x)    99.66%      |                                       [Target] Coverage (>0x)    99.66%
                                      [Target] Coverage (>=4x)    99.45%      |                                      [Target] Coverage (>=4x)    99.45%
                                     [Target] Coverage (>=10x)    98.86%      |                                     [Target] Coverage (>=10x)    98.86%
                                     [Target] Coverage (>=30x)    95.18%      |                                     [Target] Coverage (>=30x)    95.18%
                                    [Target] Coverage (>=100x)    62.56%      |                                    [Target] Coverage (>=100x)    62.56%
                                  [Target] Target Region Count    243190      |                                  [Target] Target Region Count    243190
                                  [Target] Region covered > 0x    242138      |                                  [Target] Region covered > 0x    242138
                         [Target] Fraction Region covered > 0x    99.57%      |                         [Target] Fraction Region covered > 0x    99.57%
                        [Target] Fraction Region covered >= 4x    99.29%      |                        [Target] Fraction Region covered >= 4x    99.29%
                       [Target] Fraction Region covered >= 10x    98.60%      |                       [Target] Fraction Region covered >= 10x    98.60%
                       [Target] Fraction Region covered >= 30x    94.63%      |                       [Target] Fraction Region covered >= 30x    94.63%
                      [Target] Fraction Region covered >= 100x    59.16%      |                      [Target] Fraction Region covered >= 100x    59.16%
                                            [flank] flank size    200         |                                            [flank] flank size    150
             [flank] Len of region (not include target region)    75908908    |             [flank] Len of region (not include target region)    58523561
                                         [flank] Average depth    41.41       |                                         [flank] Average depth    50.29
                                           [flank] flank Reads    49491172    |                                           [flank] flank Reads    49215762
                  [flank] Fraction of flank Reads in all reads    58.62%      |                  [flank] Fraction of flank Reads in all reads    58.29%
               [flank] Fraction of flank Reads in mapped reads    58.74%      |               [flank] Fraction of flank Reads in mapped reads    58.42%
                                        [flank] flank Data(Mb)    3143.15     |                                        [flank] flank Data(Mb)    2943.29
                    [flank] Fraction of flank Data in all data    24.87%      |                    [flank] Fraction of flank Data in all data    23.29%
                 [flank] Fraction of flank Data in mapped data    24.91%      |                 [flank] Fraction of flank Data in mapped data    23.33%
                                        [flank] Coverage (>0x)    98.38%      |                                        [flank] Coverage (>0x)    99.12%
                                       [flank] Coverage (>=4x)    92.81%      |                                       [flank] Coverage (>=4x)    96.88%
                                      [flank] Coverage (>=10x)    78.86%      |                                      [flank] Coverage (>=10x)    89.28%
                                      [flank] Coverage (>=30x)    46.34%      |                                      [flank] Coverage (>=30x)    58.22%
                                     [flank] Coverage (>=100x)    9.12%       |                                     [flank] Coverage (>=100x)    11.77%

```


# bwa 测试

缺陷：同一套数据两次跑bwa结果无法完全一样，将增加人工判定成本

优点如下：：
在不同环境中执行可保证相同结果，更利于排除因比对误差导致的一致性比较结果不一致；
重分析可增加可重复性，一致性结果利于结果解释。比如生产环境中，有需要重新分析再推送结果到报告系统，能保持前后变异结果一致；
便于后续流程对于bam的二次使用设计，如设计bam-realigment处理问题；

解决方案：
查询在不同机器不同线程数是否造成差异原因
查询是否存在随机数种子之类的参数可以让结果固定下来

调查结论：
不同线程数，获得结果不一致
不同主机相同线程数结果一致（在我们工作流中获得不一致的原因是bam文件头写入bwa路径不同导致，去掉表头再比才会相同）
-K INT 参数可以去除因线程数导致的比对误差

```bash
bwa mem -t 32 \
  -R "@RG\tID:foo\tPL:illumina\tLB:library\tSM:test" \
  /mfs1/gene/nuclear/GRCh37.fasta \
  W831-124292-9100-2022-P1_L3_1.fq.gz W831-124292-9100-2022-P1_L3_2.fq.gz | samtools sort -@ 20 -m 1024M > W831-124292-9100-2022-P1_L3.test1-1.bam

for bam in *bam ; do echo -e $bam"\t"`md5sum <(samtools view -@ 4 $bam)`; done
# W831-124292-9100-2022-P1_L3.test1-1.bam    06c0152b69ffbd3de6db90d5929cd13a    -t 32
# W831-124292-9100-2022-P1_L3.test1-2.bam    06c0152b69ffbd3de6db90d5929cd13a    -t 32
# W831-124292-9100-2022-P1_L3.test2-1.bam    556ff47763d99731e67bb4a6ac2934cf    -t 30
# W831-124292-9100-2022-P1_L3.test2-2.bam    556ff47763d99731e67bb4a6ac2934cf    -t 30
# W831-124292-9100-2022-P1_L3.test3-1.bam    04f7e6cc2945ed3519d8cf6830114fae    -t 32 -K 1000000
# W831-124292-9100-2022-P1_L3.test3-2.bam    04f7e6cc2945ed3519d8cf6830114fae    -t 32 -K 1000000
# W831-124292-9100-2022-P1_L3.test4-1.bam    04f7e6cc2945ed3519d8cf6830114fae    -t 30 -K 1000000
# W831-124292-9100-2022-P1_L3.test4-2.bam    04f7e6cc2945ed3519d8cf6830114fae    -t 30 -K 1000000

```

