
# 定量

## featureCounts

```bash
gtf="/home/gzsc/genomic/Homo_sapiens/UCSC/hg38/current/current_gtf/homo_sapiens/Homo_sapiens.GRCh38.98.chr.gtf"
featureCounts -T 12 -p -t exon -g gene_id  -a ${gtf} -o all.id.txt 01.hisat/*.bam  1>log.featureCounts 2>&1
```


featureCount 结果解读：

长度是所有区间长度之和，此处演示外显子

```python
def merge_intervals(intervals):
    result = {}
    Index = list(intervals.keys())
    (start_candidate, stop_candidate) = intervals[Index[0]]
    L_tmp_Index = [Index[0]]
    for k in Index[1:]:
        start, stop = intervals[k]
        if start <= stop_candidate:
            stop_candidate = max(stop, stop_candidate)
            L_tmp_Index.append(k)
        else:
            result[(start_candidate, stop_candidate)] = L_tmp_Index
            (start_candidate, stop_candidate) = (start, stop)
            L_tmp_Index = [k]
    result[(start_candidate, stop_candidate)] = L_tmp_Index
    return result


# 区间合并后，计算外显子长度
# all.id 结果
# ENSMUSG00000102343      1;1;1;1;1       3905739;3984225;3985160;3985160;3986147 3906134;3984298;3985984;3985351;3986215 -;-;-;-;-       1364

intervals = [
    (3905739, 3906134), (3984225, 3984298),
    (3985160, 3985984), (3985160, 3985351),
    (3986147, 3986215)
]
intervals = dict(enumerate(intervals, start=100))
l = list(merge_intervals(intervals))

# 1364
sum([x2 - x1 + 1 for x1, x2 in l])
```
